pkg/storegateway/bucket_index_metadata_fetcher.go

```diff
@@ -0,0 +1,236 @@
+		// In case a single tenant bucket index is corrupted, we don't want the store-gateway to fail at startup
+		// because unable to fetch blocks metadata. We'll act as if the tenant has no bucket index, but the query
+		// will fail anyway in the querier (the querier fails in the querier if bucket index is corrupted).
+		level.Error(f.logger).Log("msg", "corrupted bucket index found", "user", f.userID, "err", err)
+		f.metrics.synced.WithLabelValues(corruptedBucketIndex).Set(1)
+		f.metrics.submit()
+
+		return nil, nil, nil
+	}
+	if err != nil {
+		f.metrics.synced.WithLabelValues(failedMeta).Set(1)
+		f.metrics.submit()
+
+		return nil, nil, errors.Wrapf(err, "read bucket index")
+	}
+
+	// Build block metas out of the index.
+	metas = make(map[ulid.ULID]*metadata.Meta, len(idx.Blocks))
+	for _, b := range idx.Blocks {
+		metas[b.ID] = b.ThanosMeta(f.userID)
+	}
+
+	for _, filter := range f.filters {
+		var err error
+
+		// NOTE: filter can update synced metric accordingly to the reason of the exclude.
+		if customFilter, ok := filter.(MetadataFilterWithBucketIndex); ok {
+			err = customFilter.FilterWithBucketIndex(ctx, metas, idx, f.metrics.synced)
+		} else {
+			err = filter.Filter(ctx, metas, f.metrics.synced)
+		}
+
+		if err != nil {
+			return nil, nil, errors.Wrap(err, "filter metas")
+		}
+	}
+
+	for _, m := range f.modifiers {
+		// NOTE: modifier can update modified metric accordingly to the reason of the modification.
+		if err := m.Modify(ctx, metas, f.metrics.modified); err != nil {
+			return nil, nil, errors.Wrap(err, "modify metas")
+		}
+	}
+
+	f.metrics.synced.WithLabelValues(loadedMeta).Set(float64(len(metas)))
+	f.metrics.submit()
+
+	return metas, nil, nil
+}
+
+func (f *BucketIndexMetadataFetcher) UpdateOnChange(callback func([]metadata.Meta, error)) {
+	// Unused by the store-gateway.
+	callback(nil, errors.New("UpdateOnChange is unsupported"))
+}
+
+const (
+	fetcherSubSys = "blocks_meta"
+
+	corruptedMeta        = "corrupted-meta-json"
+	noMeta               = "no-meta-json"
+	loadedMeta           = "loaded"
+	failedMeta           = "failed"
+	corruptedBucketIndex = "corrupted-bucket-index"
+	noBucketIndex        = "no-bucket-index"
+
+	// Synced label values.
+	labelExcludedMeta = "label-excluded"
+	timeExcludedMeta  = "time-excluded"
+	tooFreshMeta      = "too-fresh"
+	duplicateMeta     = "duplicate"
+	// Blocks that are marked for deletion can be loaded as well. This is done to make sure that we load blocks that are meant to be deleted,
+	// but don't have a replacement block yet.
+	markedForDeletionMeta = "marked-for-deletion"
+
+	// MarkedForNoCompactionMeta is label for blocks which are loaded but also marked for no compaction. This label is also counted in `loaded` label metric.
+	MarkedForNoCompactionMeta = "marked-for-no-compact"
+
+	// Modified label values.
+	replicaRemovedMeta = "replica-label-removed"
+)
+
+// fetcherMetrics is a copy of Thanos internal fetcherMetrics. These metrics have been copied from
```

Comment:
I've opened a PR in Thanos https://github.com/thanos-io/thanos/pull/3660 to export the `fetcherMetrics` to hopefully remove the duplicated code (even if may be tricky to integrate anyway due to the additional label values `corruptedBucketIndex` and `noBucketIndex`).